### PR TITLE
Eliminate pure/elemental key words for vapor saturation related subroutines, add the vectorized versions of subroutines and change the interface for GPU computation

### PR DESCRIPTION
**Pull request by sjsprecious**
We introduce some changes to CAM codes to (1) save some computational cost and (2) make the codes portable to GPU, based on John Dennis's previous work on MG2.
- The computational cost is saved mainly due to the elimination of **pure/elemental** key words for the vapor saturation related subroutines (e.g., subroutines with name **qsat**) and using the vectorized subroutines. The major changes occur in the **wv_sat_methods.F90** and **wv_saturation.F90**, but many codes under the **physics/cam** and **chemistry** folders are also changed to accommodate the new interfaces of subroutines.
- In addition, we explicitly specify the loop index when we call the vapor saturation subroutines, which is necessary to get the code ready to be ported to GPU.
- We also make some format changes (e.g., add space) to make the coding style more consistent but this is optional.

All these changes are done based on the commit **208866b** of `cam_development` branch. We perform the following ERP test to confirm that all our changes are **BFB**.

Test suite: ERP_Ln9
Test baseline: 9-time-step CAM simulation using commit **208866b** of `cam_development` branch, making sure the tag **pumas_cam-release_v1.9** is used to check out the PUMAS Repo.
Test status: [BFB]

### CODE REVIEW

File: src/physics/cam/clubb_intr.F90 [hunk 1]
```diff
@@ -1893,11 +1893,14 @@ real(r8) :: inv_exner_clubb(pcols,pverp)     ! Inverse exner function consistent
...
      initend(:ncol,:)=0._r8
 
-     call ice_macro_tend(naai(:ncol,top_lev:pver),state1%t(:ncol,top_lev:pver), &
-        state1%pmid(:ncol,top_lev:pver),state1%q(:ncol,top_lev:pver,1),state1%q(:ncol,top_lev:pver,ixcldice),&
-        state1%q(:ncol,top_lev:pver,ixnumice),latsub,hdtime,&
-        stend(:ncol,top_lev:pver),qvtend(:ncol,top_lev:pver),qitend(:ncol,top_lev:pver),&
-        initend(:ncol,top_lev:pver))
+     do k = top_lev, pver
+        do i = 1, ncol
+           call ice_macro_tend(naai(i,k),state1%t(i,k),state1%pmid(i,k),      &
```
Comment (johnmauff) [inline src/physics/cam/clubb_intr.F90]: Do we want to create a vectorized version of ice_macro_tend?

File: src/physics/cam/clubb_intr.F90 [hunk 2]
```diff
@@ -2796,12 +2799,15 @@ real(r8) :: inv_exner_clubb(pcols,pverp)     ! Inverse exner function consistent
...
  
-      call liquid_macro_tend(npccn(:ncol,top_lev:pver),state1%t(:ncol,top_lev:pver), &
-         state1%pmid(:ncol,top_lev:pver),state1%q(:ncol,top_lev:pver,ixq),state1%q(:ncol,top_lev:pver,ixcldliq),&
-         state1%q(:ncol,top_lev:pver,ixnumliq),latvap,hdtime,&
-         stend(:ncol,top_lev:pver),qvtend(:ncol,top_lev:pver),qctend(:ncol,top_lev:pver),&
-         inctend(:ncol,top_lev:pver))
- 
+      do k = top_lev, pver
+         do i = 1, ncol
+            call liquid_macro_tend(npccn(i,k),state1%t(i,k), state1%pmid(i,k),    &
```
Comment (johnmauff) [inline src/physics/cam/clubb_intr.F90]: Do we want to create a vectorized version of liquid_macro_tend?

File: src/physics/cam/wv_saturation.F90 [hunk 3]
```diff
@@ -259,41 +272,41 @@ end subroutine wv_sat_final
...
+  real(r8)              :: es ! SVP (Pa)
 
-  es = wv_sat_svp_water(T)
+  es = wv_sat_svp_water(t)
 
 end function svp_water
 
 ! Compute saturation vapor pressure over ice
-elemental function svp_ice(t) result(es)
+function svp_ice(t) result(es)
```
Comment (johnmauff) [inline src/physics/cam/wv_saturation.F90:305]: Probably want to create a vectorized version of svp_ice

File: src/physics/cam/wv_saturation.F90 [hunk 4]
```diff
@@ -259,41 +272,41 @@ end subroutine wv_sat_final
 !---------------------------------------------------------------------
 
 ! Compute saturation vapor pressure over water
-elemental function svp_water(t) result(es)
+function svp_water(t) result(es)
```
Comment (johnmauff) [inline src/physics/cam/wv_saturation.F90:278]: Create a vectorized version of svp_water.

File: src/physics/cam/wv_saturation.F90 [hunk 5]
```diff
@@ -259,41 +272,41 @@ end subroutine wv_sat_final
...
+  real(r8)              :: es ! SVP (Pa)
 
-  es = wv_sat_svp_ice(T)
+  es = wv_sat_svp_ice(t)
 
 end function svp_ice
 
 ! Compute saturation vapor pressure with an ice-water transition
-elemental function svp_trans(t) result(es)
+function svp_trans(t) result(es)
```
Comment (johnmauff) [inline src/physics/cam/wv_saturation.F90:332]: Create a vectorized version of svp_trans

File: src/physics/cam/clubb_intr.F90 (hunk 1 above)
Comment (sjsprecious) [reply src/physics/cam/clubb_intr.F90]: Thanks John. Just push the vectorized version of ice_macro_tend.

File: src/physics/cam/clubb_intr.F90 (hunk 2 above)
Comment (sjsprecious) [reply src/physics/cam/clubb_intr.F90]: Thanks John. Just push the vectorized version of liquid_macro_tend.
The liquid_macro_tend and ice_macro_tend belong to CLUBB so I am not sure whether we should change them in our MG3 work.

File: src/physics/cam/wv_saturation.F90 (hunk 5 above)
Comment (sjsprecious) [reply src/physics/cam/wv_saturation.F90:332]: Thanks, it is added.

File: src/physics/cam/wv_saturation.F90 (hunk 4 above)
Comment (sjsprecious) [reply src/physics/cam/wv_saturation.F90:278]: Thanks, it is added.

File: src/physics/cam/wv_saturation.F90 (hunk 3 above)
Comment (sjsprecious) [reply src/physics/cam/wv_saturation.F90:305]: Thanks, it is added.

**Review by Katetc — APPROVED** [review]

**Comment by Katetc** [comment]
Just a note that I changed the base for this PR from PUMAS_Development/CAM/cam_development (which we keep as a mirror to ESCOMP/CAM/cam_development) to a living branch I just made, cam_pumas_pr and will use to stage the ESCOMP/CAM/cam_development pull request. I will tag @sjsprecious in the ESCOMP/CAM PR so you can see the CAM SE review suggestions, but it is unlikely you will need to do any further work. We are also going to fold in a few other issues and code changes to this CAM PR while we have the tag, and I will need this living branch for that.

**Pull request merged**